### PR TITLE
feat(rest): add mountExpressRouter

### PR DESCRIPTION
**Pull request by nabdelgadir**
Closes https://github.com/strongloop/loopback-next/issues/2389.

Add `mountExpressRouter` to `RestApplication` and `RestServer` and include in `Routes.md`.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/rest/src/router/external-express-routes.ts [hunk 1]
```diff
@@ -42,6 +50,17 @@ export class ExternalExpressRoutes {
...
+    router: ExpressRequestHandler,
+    spec: RouterSpec = {paths: {}},
+  ) {
+    this._externalRoutes.use(basePath, router);
+
+    spec = rebaseOpenApiSpec(spec, basePath);
+    assignRouterSpec(this._specForExternalRoutes, spec);
+  }
+
   find(request: Request): ResolvedRoute {
```
Comment (jannyHou) [inline packages/rest/src/router/external-express-routes.ts]: @nabdelgadir When we was debugging it I told you the problem is raised by undefined `spec` while that's not true, don't be mislead by that wrong assersion :) I think the real problem is this:

For the endpoint your are testing `/dog/pug`, the route lookup happens in the following way:
1. in your app/server's sequence: `const route = this.findRoute(request);`
2. goes to the http handler's [`findRoute` function](https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/http-handler.ts#L61-L64)
3. calls the find method in routing table and falls to the [external route condition](https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/router/routing-table.ts#L138-L145)
4. finally reaches the `find` method at line#64 here

Now if you look at the implementation of the function `find`, it always returns an external route instanciated with `this._staticRoutes`, WHILE your endpoint is registered by calling `mountRouter` method, which calls `this._externalRoutes.user(basePath, router)` to register the router/handler.

To make a non-static external route work, I believe you need to do some change in the `find` method and also the routing table file:

- IMO an external route is NOT equal to a static route, but [our code](https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/router/routing-table.ts#L138-L145) assumes it's always static, this could be told from the debug message and also the `find` method here

- Improve the `find` method here to distinguish between static routes and external(non-static) routes, so that `/dog/pug` instantiate `ExternalRoute` with router `_externalRoutes`, instead of `_staticRoutes`

cc @bajtos @hacksparrow may know more about the routing stuff, ^ correct me if wrong :) thanks.

Comment (jannyHou) [reply packages/rest/src/router/external-express-routes.ts]: quickly hacked code to 
```ts

  find(request: Request): ResolvedRoute {
    // NOT a solution, just a hack
    return new ExternalRoute(
      this._externalRoutes,
      request.method,
      request.url,
      {
        description: 'LoopBack static assets route',
        'x-visibility': 'undocumented',
        responses: {},
      },
    );
  }
```
And your test pass. So I think your tests are written in the correct way(create a new express router then register routes to it)

Comment (bajtos) [reply packages/rest/src/router/external-express-routes.ts]: I don't have time to look into this in details today. If you look at my PoC in https://github.com/strongloop/loopback-next/pull/2318, you can see that the resolved route is accepting both static and external routes.

https://github.com/strongloop/loopback-next/blob/379e0c050c49e70713cabc9aa6d85a892250d375/packages/rest/src/router/external-express-routes.ts#L85-L91

When that route is invoked, it calls the external router to handle the request fist, and if the request is not handled, then it calls static assets router second.

https://github.com/strongloop/loopback-next/blob/379e0c050c49e70713cabc9aa6d85a892250d375/packages/rest/src/router/external-express-routes.ts#L102-L114

Comment (bajtos) [reply packages/rest/src/router/external-express-routes.ts]: > Improve the find method here to distinguish between static routes and external(non-static) routes, so that /dog/pug instantiate ExternalRoute with router _externalRoutes, instead of _staticRoutes

I don't think it's feasible to tell at `findRoute` time whether the request will match an external route or a static-assets route.

Comment (jannyHou) [reply packages/rest/src/router/external-express-routes.ts]: > When that route is invoked, it calls the external router to handle the request fist, and if the request is not handled, then it calls static assets router second.

> I don't think it's feasible to tell at findRoute time whether the request will match an external route or a static-assets route.

Thanks! Got it.

File: docs/site/Routes.md [hunk 2]
```diff
@@ -226,3 +226,40 @@ export class MyApplication extends RestApplication {
...
+import {RestApplication, RouterSpec} from '@loopback/rest';
+import {Request, Response} from 'express';
+import * as express from 'express';
+
+const app = new RestApplication();
+const router = express.Router();
+
+router.get('/pug', function(_req: Request, res: Response) {
+  res.send('Pug!');
+});
```
Comment (bajtos) [inline docs/site/Routes.md]: Since our primary motivation for enabling this feature is to allow existing Express applications to be mounted in LB4, I would prefer if this example was more Express-app centric.

```ts
const legacyApp = express();
// your existing Express routes
legacyApp.get('/pug', function(_req: Request, res: Response) {
  res.send('Pug!');
});

// compose them with the new LB4 app
const spec: RouterSpec = {
  // insert your spec here, your 'paths', 'components', and 'tags' will be used
};
	
app.mountExpressRouter('/dogs', legacyApp, spec);
```

Also, I think `mountExpressRouter` should be typically called from the application constructor. To promote this best practice, I would prefer the example to use the same approach too.

```ts
const legacyApp = require('./express-app');

const legacySpec = {
   // insert your spec here, your 'paths', 'components', and 'tags' will be used
}

class MyApplication extends RestApplication {
  constructor(/*...*/) {
    // ...
   
    this.mountExpressRouter('/dogs', legacyApp, legacySpec);
  }
}
```

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 3]
```diff
@@ -5,9 +5,12 @@
 
 import {anOperationSpec} from '@loopback/openapi-spec-builder';
 import {Client, createRestAppClient, expect} from '@loopback/testlab';
+import * as express from 'express';
+import {Request, Response} from 'express';
 import * as fs from 'fs';
 import * as path from 'path';
-import {RestApplication, RestServer, RestServerConfig, get} from '../..';
+import {get, RestApplication, RestServer, RestServerConfig} from '../..';
+import {RouterSpec} from '../../router/router-spec';
```
Comment (bajtos) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: Can we import `RouterSpec` from `../..`, similarly as we do for other types?

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 4]
```diff
@@ -163,6 +166,73 @@ describe('RestApplication (integration)', () => {
     await client.get(response.header.location).expect(200, 'Hi');
   });
 
+  context('mounting an express router on a LoopBack application', async () => {
+    givenApplication();
+
+    const router = express.Router();
+    await restApp.start();
+    client = createRestAppClient(restApp);
```
Comment (bajtos) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: Test setup belongs to a `before` or `beforeEach` hook.

That way the setup code is run only when the tests inside this `context` block are actually executed (think of `it.only` used elsewhere). More importantly, when the setup is executed from a hook, any errors are correctly associated with the relevant test case(s).

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 5]
```diff
@@ -163,6 +166,73 @@ describe('RestApplication (integration)', () => {
...
+      router.get('/pug', function(_req: Request, res: Response) {
+        res.send('Pug!');
+      });
+      restApp.mountExpressRouter('/dogs', router);
+
+      await client.get('/dogs/poodle/').expect(200, 'Poodle!');
+      await client.get('/dogs/pug').expect(200, 'Pug!');
+    });
+
+    it.skip('mounts an express Router with spec with paths', async () => {
```
Comment (bajtos) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: `it.skip` - I assume this is not implemented yet?

File: packages/rest/src/rest.application.ts [hunk 6]
```diff
@@ -264,4 +270,22 @@ export class RestApplication extends Application implements HttpServerLike {
...
+
+  /**
+   * Mount an Express router to expose additional REST endpoints handled
+   * via legacy Express-based stack.
+   *
+   * @param basePath Path where to mount the router at, e.g. `/` or `/api`.
+   * @param router The Express router to handle the requests.
+   * @param spec A partial OpenAPI spec describing endpoints provided by the router.
+   * LoopBack will prepend `basePath` to all endpoints automatically. Use `undefined`
+   * if you don't want to document the routes.
```
Comment (bajtos) [inline packages/rest/src/rest.application.ts]:
```diff
- Use `undefined` if you don't want to document the routes.
+ This argument is optional. You can leave it out if you don't want to document the routes.
```

File: packages/rest/src/rest.server.ts [hunk 7]
```diff
@@ -827,6 +829,24 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
+
+  /**
+   * Mount an Express router to expose additional REST endpoints handled
+   * via legacy Express-based stack.
+   *
+   * @param basePath Path where to mount the router at, e.g. `/` or `/api`.
+   * @param router The Express router to handle the requests.
+   * @param spec A partial OpenAPI spec describing endpoints provided by the router.
+   * LoopBack will prepend `basePath` to all endpoints automatically. Use `undefined`
+   * if you don't want to document the routes.
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]:
```diff
- Use `undefined` if you don't want to document the routes.
+ This argument is optional. You can leave it out if you don't want to document the routes.
```

File: packages/rest/src/router/external-express-routes.ts [hunk 8]
```diff
@@ -42,12 +53,29 @@ export class ExternalExpressRoutes {
...
-      'x-visibility': 'undocumented',
-      responses: {},
-    });
+    return new ExternalRoute(
+      this._externalRoutes,
+      this._staticRoutes,
+      request.method,
+      request.url,
+      {
+        description: 'LoopBack static assets route',
```
Comment (bajtos) [inline packages/rest/src/router/external-express-routes.ts]: This description is no longer true, please update it. E.g.

```suggestion
        description: 'External route or a static asset',
```

File: packages/rest/src/router/external-express-routes.ts [hunk 9]
```diff
@@ -90,6 +126,24 @@ class ExternalRoute implements RouteEntry, ResolvedRoute {
   }
 }
 
+export function rebaseOpenApiSpec<T extends Partial<OpenApiSpec>>(
```
Comment (bajtos) [inline packages/rest/src/router/external-express-routes.ts]: Please add few unit-level tests to verify how `rebaseOpenApiSpec` handles different edge cases. Take a look at `if` conditions below to see what to test.

File: packages/rest/src/router/router-spec.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {OpenApiSpec} from '@loopback/openapi-v3-types';
+
+export type RouterSpec = Pick<OpenApiSpec, 'paths' | 'components' | 'tags'>;
+
+export function assignRouterSpec(target: RouterSpec, additions: RouterSpec) {
```
Comment (bajtos) [inline packages/rest/src/router/router-spec.ts]: Similarly here. Please add few unit-level tests to verify how `assignRouteSpec` handles different edge cases. Take a look at `if` conditions and explanatory code comments below to see what to test.

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 11]
```diff
@@ -163,6 +166,73 @@ describe('RestApplication (integration)', () => {
     await client.get(response.header.location).expect(200, 'Hi');
   });
 
+  context('mounting an express router on a LoopBack application', async () => {
```
Comment (bajtos) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: Please add a test to verify the order of precedence of external vs. static-asset routes.

```ts
app.staticAssets('/test', /* path to some dir */);
app.mountExpressRouter('/test', /* a router */);
// ...

await client.get(
 // a URL matching a static asset and an Express route too
).expect(
  // the express route was invoked
);
```

File: docs/site/Routes.md [hunk 12]
```diff
@@ -226,3 +226,57 @@ export class MyApplication extends RestApplication {
   }
 }
 ```
+
+## Mounting an Express Router
+
+If you have an existing [Express](https://expressjs.com/) application that you
+want to use with LoopBack 4, you can mount the Express application on top of a
+LoopBack 4 application. This way you can mix and match both frameworks, while
+using LoopBack as the host. Mounting an Express router on a LoopBack 4
```
Comment (bajtos) [inline docs/site/Routes.md]: Lovely 👏 

Can we also add a link to docs explaining how to compose LB4 and Express the other way, where Express is acting as the host and LB4 routes are mounted?

File: docs/site/Routes.md [hunk 13]
```diff
@@ -226,3 +226,57 @@ export class MyApplication extends RestApplication {
...
+import * as express from 'express';
+
+const legacyApp = express();
+
+// your existing Express routes
+legacyApp.get('/pug', function(_req: Request, res: Response) {
+  res.send('Pug!');
+});
+
+module.exports = legacyApp;
```
Comment (bajtos) [inline docs/site/Routes.md]: Since we are TypeScript, would the following work too?

```suggestion
export {legacyApp};
```
```

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 14]
```diff
@@ -163,6 +171,90 @@ describe('RestApplication (integration)', () => {
...
+                },
+              },
+            },
+          },
+        },
+      };
+      router.get('/hello', greetDogs);
+      restApp.mountExpressRouter('/dogs', router, spec);
+      await client.get('/dogs/hello').expect(200, 'Hello dogs!');
+    });
```
Comment (bajtos) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: In this test, it's important to verify that `GET /hello` endpoint is included in the spec returned by `restApp.restServer.getApiSpec()`.

File: packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  const target: RouterSpec = {paths: {}, components: {}, tags: []};
+  const additions: RouterSpec = {paths: {}, components: {}, tags: []};
+
+  it('duplicates the additions spec if the target spec is empty', async () => {});
+
+  it('does not assign components without schema', async () => {});
+
+  it('uses the route registered first', async () => {});
+
+  it('does not duplicate tags from the additional spec', async () => {});
```
Comment (bajtos) [inline packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts]: Nice list of test cases :+1:

File: packages/rest/src/router/routing-table.ts [hunk 16]
```diff
@@ -137,7 +137,7 @@ export class RoutingTable {
 
     if (this._externalRoutes) {
       debug(
-        'No API route found for %s %s, trying to find a static asset',
+        'No API route found for %s %s, trying to find external Express route',
```
Comment (bajtos) [inline packages/rest/src/router/routing-table.ts]:
```suggestion
        'No API route found for %s %s, trying to find an external Express route',
```

File: docs/site/Routes.md [hunk 17]
```diff
@@ -226,3 +226,62 @@ export class MyApplication extends RestApplication {
...
 ```
+
+## Mounting an Express Router
+
+If you have an existing [Express](https://expressjs.com/) application that you
+want to use with LoopBack 4, you can mount the Express application on top of a
+LoopBack 4 application. This way you can mix and match both frameworks, while
+using LoopBack as the host. You can also do the opposite and use Express as the
+host by mounting LoopBack 4 REST API on an Express application. See
+[Creating an Express Application with LoopBack REST API](express-with-lb4-rest-tutorial.md)
```
Comment (bajtos) [inline docs/site/Routes.md]: Should we update `express-with-lb4-rest-tutorial.md` to link back to this section too?

File: docs/site/Routes.md [hunk 18]
```diff
@@ -226,3 +226,62 @@ export class MyApplication extends RestApplication {
...
+for the tutorial.
+
+Mounting an Express router on a LoopBack 4 application can be done using the
+`mountExpressRouter` function provided by both `RestApplication` and
+`RestServer`.
```
Comment (bajtos) [inline docs/site/Routes.md]: How about replacing naked class names with links to API docs?

- http://apidocs.loopback.io/@loopback%2fdocs/rest.html#RestApplication
- http://apidocs.loopback.io/@loopback%2fdocs/rest.html#RestServer

File: packages/rest/src/__tests__/unit/router/rebase-openapi-spec.unit.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+              },
+            },
+          },
+        },
+      },
+    };
+
+    rebaseOpenApiSpec(spec, '/greetings');
+
+    expect(spec).to.equal(spec);
```
Comment (bajtos) [inline packages/rest/src/__tests__/unit/router/rebase-openapi-spec.unit.ts]: This is a [Tautology](https://en.wikipedia.org/wiki/Tautology_(logic)) - you are asserting that `spec` is a pointer to the same object as pointed to by `spec`.

I think you should check that `spec.paths` has entries `/` and `/hello` after the rebase.

```suggestion
    expect(Object.keys(spec.paths)).to.deepEqual('/', '/hello');
```

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 20]
```diff
@@ -163,6 +171,103 @@ describe('RestApplication (integration)', () => {
...
+      });
+
+      restApp.static('/dogs', ASSETS);
+      restApp.mountExpressRouter('/dogs', router);
+
+      await client.get('/dogs/').expect(200, 'External dog');
+    });
+
+    it('mounts an express Router without spec', async () => {
+      const router = express.Router();
```
Comment (b-admike) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: What would the default generated(?) spec look like here if one is not provided? Can we add an assertion for it?

File: packages/rest/src/__tests__/integration/rest.application.integration.ts [hunk 21]
```diff
@@ -163,6 +171,103 @@ describe('RestApplication (integration)', () => {
...
+      client = createRestAppClient(restApp);
+    });
+
+    it('gives precedence to an external route over a static route', async () => {
+      const router = express.Router();
+      router.get('/', function(_req: Request, res: Response) {
+        res.send('External dog');
+      });
+
+      restApp.static('/dogs', ASSETS);
```
Comment (b-admike) [inline packages/rest/src/__tests__/integration/rest.application.integration.ts]: Do we silently replace the static route in this case?

File: packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    };
+
+    assignRouterSpec(target, additions);
```
Comment (b-admike) [inline packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts]: IIUC, we can make this assertion a little more clearer for others:
```suggestion
    const resultSpec = assignRouterSpec(target, additions);
    expect(resultSpec.paths).to.equal(target.paths);
```

File: packages/rest/src/__tests__/unit/router/rebase-openapi-spec.unit.ts [hunk 23]
```diff
@@ -0,0 +1,144 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {rebaseOpenApiSpec} from '../../../router';
+
+describe('rebaseOpenApiSpec', () => {
+  it('does not modify the OpenApiSpec if it does not have paths', async () => {
```
Comment (b-admike) [inline packages/rest/src/__tests__/unit/router/rebase-openapi-spec.unit.ts]: Nitpick:
```suggestion
  it('does not modify an OpenAPI spec if it does not have paths', async () => {
```

File: packages/rest/src/__tests__/unit/router/rebase-openapi-spec.unit.ts [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+                description: 'OK',
+              },
+            },
+          },
+        },
+      },
+    };
+
+    let rebasedSpec = rebaseOpenApiSpec(spec, '');
+    expect(spec).to.eql(rebasedSpec);
```
Comment (b-admike) [inline packages/rest/src/__tests__/unit/router/rebase-openapi-spec.unit.ts]: nitpick (we usually expect the result to match some expected value):
```suggestion
    expect(rebasedSpec).to.eql(spec);
```

File: packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts (hunk 22 above)
Comment (nabdelgadir) [reply packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts]: `assignRouterSpec` has a `void` return type, but I changed `expect(target.paths).to.equal(target.paths);` to `expect(target.paths).to.eql(target.paths);`

File: packages/rest/src/__tests__/integration/rest.application.integration.ts (hunk 20 above)
Comment (nabdelgadir) [reply packages/rest/src/__tests__/integration/rest.application.integration.ts]: The `spec` is optional, so if it isn't provided then it won't get generated or added to `openapi.json`.

File: packages/rest/src/__tests__/integration/rest.application.integration.ts (hunk 21 above)
Comment (nabdelgadir) [reply packages/rest/src/__tests__/integration/rest.application.integration.ts]: When it's looking for the route, it first checks if there's an external route, then if it can't find it then it uses the static one (see [here](https://github.com/strongloop/loopback-next/pull/2643/files#diff-4ac1e4b2069b2ff62309cfdd6b9dbe7bR103)), so I believe the static one still exists, but is overshadowed by the external one.

File: docs/site/Routes.md [hunk 25]
```diff
@@ -226,3 +226,64 @@ export class MyApplication extends RestApplication {
...
+```
+
+{% include code-caption.html content="src/application.ts" %}
+
+```ts
+import {RestApplication} from '@loopback/rest';
+
+const legacyApp = require('./express-app').legacyApp;
+
+const legacySpec: RouterSpec = {
```
Comment (raymondfeng) [inline docs/site/Routes.md]: The variable name `legacySpec` is confusing. It should be something like `openApiSpecForLegacyApp`.

File: packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts [hunk 26]
```diff
@@ -0,0 +1,151 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {assignRouterSpec, RouterSpec} from '../../../router';
```
Comment (raymondfeng) [inline packages/rest/src/__tests__/unit/router/assign-router-spec.unit.ts]: It's better to import from `../../..` (the `@loopback/rest` module).

File: packages/rest/src/rest.application.ts [hunk 27]
```diff
@@ -264,4 +270,23 @@ export class RestApplication extends Application implements HttpServerLike {
...
+   *
+   * @param basePath Path where to mount the router at, e.g. `/` or `/api`.
+   * @param router The Express router to handle the requests.
+   * @param spec A partial OpenAPI spec describing endpoints provided by the
+   * router. LoopBack will prepend `basePath` to all endpoints automatically.
+   * This argument is optional. You can leave it out if you don't want to
+   * document the routes.
+   */
+  mountExpressRouter(
+    basePath: string,
```
Comment (raymondfeng) [inline packages/rest/src/rest.application.ts]: Do we allow `RegExp`?

Comment (nabdelgadir) [reply packages/rest/src/rest.application.ts]: An express router can take in `RegExp` so it's possible, but I'll do this in a different PR so this one can be landed and I can also add tests.

Comment (raymondfeng) [reply packages/rest/src/rest.application.ts]: Maybe it's not practical as OpenAPI spec does not support regexp/wildcard for `path`.
